### DISCUSSION TITLE
Proposal: "while" as a LINQ query expression clause

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#94] [General]
Copied from https://github.com/dotnet/roslyn/issues/15638

Low effort/reward proposal:

I propose that the existing C# keyword `while` be added as a LINQ query expression clause.  Unlike `where` which filters out elements that do not meet the condition `while` would include all elements as long as the condition is met.  Since `while` is already a keyword this would not break any existing syntax.

```cs
var q = from i in numbers
where i % 2 == 0
while i < 10
select i * 2;

// translated into
var q = numbers
    .Where(i => i % 2 == 0)
    .TakeWhile(i => i < 10)
    .Select(i => i * 2);
```

**Comment by orthoxerox** [comment]
Should we have an issue for each new LINQ clause, or will you collect them all here?

**Comment by rubberduck203** [comment]
What does `while i < 10` accomplish that `where i < 10` doesn't? Do they accomplish the same thing, or is there a semantic difference?

**Comment by jnm2** [comment]
@ckuhn203 [While](https://msdn.microsoft.com/en-us/library/system.linq.enumerable.takewhile.aspx) skips the rest of the items as soon as any item doesn't match. [Where](https://msdn.microsoft.com/en-us/library/system.linq.enumerable.where.aspx) only skips the items that don't match.
`0,1,2 while i % 2 == 0` gives `0`.
`0,1,2 where i % 2 == 0` gives `0,2`.

**Comment by HaloFour** [comment]
@orthoxerox I'd think that each would be managed separately (although if you'd be suggesting a family of clauses I could see them proposed as one).  My thoughts for this particular proposal was that it's relatively low hanging fruit in that it's syntactically unambiguous given that `while` is a reserved keyword and it maps easily to the existing `TakeWhile` extension method.

**Comment by orthoxerox** [comment]
@HaloFour have you thought about extending this to `SkipWhile()`?

**Comment by vladd** [comment]
Is this new operator easily translatable to SQL?

(And I'm afraid it's not easily parallelizable, as it depends on the element ordering.)

**Comment by jnm2** [comment]
@Vladd it would be possible but complicated to translate to SQL. However, there's no requirement for any ORM to implement support for this. For example, Entity Framework has never supported `First`, `Single`, `SingleOrDefault`, etc, in a subquery.

**Comment by orthoxerox** [comment]
@vladd It's an existing LINQ method, `TakeWhile`. Adding it to query syntax doesn't affect LINQ providers in any way.
